Create a single table for differentially expressed genes

delfiterradas opened the pull request:
<!--
# nf-core/differentialabundance pull request

Many thanks for contributing to nf-core/differentialabundance!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/differentialabundance/tree/master/.github/CONTRIBUTING.md)
-->
Closes https://github.com/nf-core/differentialabundance/issues/574
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/differentialabundance/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/differentialabundance _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

antoniasaracco committed:
Create a single table for differentially expressed genes instead of spliting them by Direction

antoniasaracco committed:
Update changelog

delfiterradas committed:
Fix whitespace

delfiterradas commented:
This is the current report: 
[SRP254919_differentialabundance_report_current.html](https://github.com/user-attachments/files/24173417/SRP254919_differentialabundance_report_current.html)

And this is what the new version would look like with a single table for differentially expressed genes:
[SRP254919_differentialabundance_report_new.html](https://github.com/user-attachments/files/24173430/SRP254919_differentialabundance_report_new.html)

delfiterradas committed:
Update changelog

grst commented on assets/differentialabundance_report.qmd:
```diff
@@ -983,56 +990,100 @@ for (i in 1:nrow(contrasts)) {
+
+        if (length(parts) > 0) {
+            # Bind rows even if columns differ slightly
+            combined <- do.call(rbind, parts)
```
Seems like we are splitting the table somewhere and then concatenating it again. 

There's also a table `full_de` above, why can't we reuse it here? Adding a `Direction` column should then be a simple `mutate()`.

grst commented on assets/differentialabundance_report.qmd:
```diff
@@ -983,56 +990,100 @@ for (i in 1:nrow(contrasts)) {
+            cols_present <- intersect(cols_to_round, colnames(combined))
+            if (length(cols_present)) {
+                combined[, cols_present] <- signif(combined[, cols_present, drop = FALSE], 8)
+            }
```
I think a more elegant solution than rounding the columns would be to use formatting verbs for `datatable`: 

```r
datatable(df) |> formatRound(
  c(params$meta$params$differential_fc_column, params$meta$params$differential_pval_column),
  5
)
```

atrigila commented:
As agreed in today's meeting this is owned by @antoniasaracco now

Commit:
replace table splitting + fix formatting

Commit:
Refactor differential genes table to use full_de directly and formatRound()

- Use full_de filtered by de_fc & de_pval instead of sig_differential
- Add Direction column based on fold-change sign
- Replace signif() with formatRound() for display formatting
- Addresses reviewer feedback from PR #580

antoniasaracco committed:
run pre-commit run --all-files

antoniasaracco commented:
hi @grst, I've addressed your comments and it's ready for review. Thanks!

grst commented on assets/differentialabundance_report.qmd:
```diff
@@ -983,56 +990,97 @@ for (i in 1:nrow(contrasts)) {
+            first_cols <- c("Direction", setdiff(colnames(combined), "Direction"))
+            combined <- combined[, first_cols, drop = FALSE]
+            colnames(combined) <- prettifyVariablename(colnames(combined))
+            combined <- round_dataframe_columns(combined, digits = params$meta$params$round_digits)
```
I suggest rewriting this with dplyr, basically

```
combined = full_de |> 
 mutate(direction = ...) |>
 relocate(Direction) |>
 rename_with(prettifyVariablename) |>
 round_dataframe_columns()
```
 
Do we even need `round_dataframe_columns`, given the `formatRound` below?

grst replied:
I noticed dplyr is not a dependency yet. I think we should still go that way. Actually I'm surprised we managed to get to 1000+ lines without adding dplyr. 

We should gradually make more use of it in this report - it's much more legible and can probably save us quite some lines of code in the long run.

antoniasaracco replied:
hi @grst I have updated the code to implement dplyr library. Please let me know your thoughts, thanks!

antoniasaracco committed:
Implement dplyr library

antoniasaracco committed:
Merge branch 'dev' into single_table

antoniasaracco committed:
Fix figure size

antoniasaracco committed:
Merge branch 'single_table' of https://github.com/delfiterradas/differentialabundance into single_table

pinin4fjords commented:
I'm seeing some issues in the report, but I think it's because of out-of-date examples. Could you supply fresh ones please?

antoniasaracco commented:
@pinin4fjords 
[SRP254919_differentialabundance_report_updated.html](https://github.com/user-attachments/files/24494805/SRP254919_differentialabundance_report_updated.html)

pinin4fjords commented:
<img width="780" height="967" alt="old" src="https://github.com/user-attachments/assets/089ca5f3-5f6c-493f-93fb-0cde029b0929" />
<img width="1078" height="1006" alt="new" src="https://github.com/user-attachments/assets/05588f1c-f324-435a-ac1c-cb55bb730d11" />

What happened to the thresholds in the plots?

Also, I'm not convinced by the 'differential status' column- isn't it going to be the same (except for the sign of the fold change) in all rows? Seems like a waste of screen real estate.

pinin4fjords commented:
I'm also not entirely in agreement with https://github.com/nf-core/differentialabundance/issues/491. I can live with this, but only if we have a way of easily filtering the tables to up- and down- only. I'm pretty sure `DT` could give us the controls to do that- could you add them please?

grst commented:
Well, we are on it, but something seems broken in `DT`: https://github.com/nf-core/differentialabundance/issues/518

What works for now is to just type "up" or "down" in the search bar, or just sort by logFC

pinin4fjords commented:
OK, fair enough (points still stand on the threshold etc)

Commit:
Fix missing threshold

antoniasaracco committed:
Merge branch 'dev' into single_table

antoniasaracco commented:
Hi @pinin4fjords,

I have the missing threshold issue. @grst what is your opinion on the comment 'differential status' column?

Thanks!

grst commented:
> @grst what is your opinion on the comment 'differential status' column?

I agree it's unnecessary. 

Could you please post screenshots or HTML report of the latest version?

pinin4fjords commented on assets/differentialabundance_report.qmd:
```diff
@@ -998,19 +1003,19 @@ for (i in 1:nrow(contrasts)) {
         full_de$differential_status <- factor(full_de$differential_status, levels = c("Not significant", de_fc_label, de_pval_label, de_pval_fc_label), ordered = TRUE) # Factorize status so that non-significant is always first
         # Define the thresholds we'll draw
 
+        # OLD (working)
```
Don't leave this in

Commit:
remove comment

Commit:
Merge branch 'single_table' of https://github.com/delfiterradas/differentialabundance into single_table

antoniasaracco committed:
Merge branch 'dev' into single_table

Commit:
remove differential_status column

Commit:
Merge branch 'single_table' of https://github.com/delfiterradas/differentialabundance into single_table

antoniasaracco commented:
@grst, @pinin4fjords Here is the update report. Please let me know what you think!
[SRP254919_differentialabundance_report_new.html](https://github.com/user-attachments/files/24721517/SRP254919_differentialabundance_report_new.html)

pinin4fjords commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- [[#491]](https://github.com/nf-core/differentialabundance/pull/509) - Do not split results table into up- and downregulated. ([@antoniasaracco] (https://github.com/antoniasaracco), review by [@atrigila](https://github.com/atrigila), [@grst](https://github.com/grst) and [@pinin4fjords](https://github.com/pinin4fjords)).
```
```suggestion
- [[#580]](https://github.com/nf-core/differentialabundance/pull/580) - Do not split results table into up- and downregulated. ([@antoniasaracco](https://github.com/antoniasaracco), review by [@atrigila](https://github.com/atrigila), [@grst](https://github.com/grst) and [@pinin4fjords](https://github.com/pinin4fjords)).
```

pinin4fjords reviewed (approved):
Changelog needs a fix, otherwise good

grst reviewed (dismissed):
the horizontal threshold line still looks weird for me in the latest file you shared. 
IMO it should span the entire plot

<img width="1601" height="717" alt="image" src="https://github.com/user-attachments/assets/5adade58-2840-4a61-9dda-4059351b51cf" />

grst committed:
Update CHANGELOG.md

Co-authored-by: Jonathan Manning <pininforthefjords@gmail.com>

pinin4fjords commented:
> the horizontal threshold line still looks weird for me in the latest file you shared. IMO it should span the entire plot
> 


That's how it was before I _think_. It spans the range of points. Though that could be improved obviously.

antoniasaracco commented:
@grst Is like this okay? I had to add it manually after the plot since in the plot we are setting the x axis range
```
p <- do.call(plotly_scatterplot, plot_args) %>%
            layout(xaxis = list(range = list(-max_fc, max_fc))) %>%
            add_segments(
                x = -max_fc, 
                xend = max_fc,
                y = -log10(p_value_thresholds[[pvt]]), 
                yend = -log10(p_value_thresholds[[pvt]]),
                line = list(color = "black", width = 2, dash = "dot"),
                showlegend = TRUE,
                inherit = FALSE,
                name = paste(pval_column, "=", p_value_thresholds[[pvt]])
            )
```
<img width="743" height="308" alt="image" src="https://github.com/user-attachments/assets/7feb4164-ecae-40c3-90b2-0c59d6fa9370" />

antoniasaracco committed:
Merge branch 'dev' into single_table

grst commented:
Hmm, the proper solution would be to fix this in shinyngs. We have a couple of options
 1. apply the workaround proposed by @antoniasaracco above
 2. fix this in shinyngs
 3. rewrite the volcano plot without shinyngs. Either in pure plotly, or by combining Bioconductor's EnhancedVolcano library with ggploty as described in https://github.com/kevinblighe/EnhancedVolcano/issues/114

For anything except (1), I would prefer to decouple it from this PR and create a separate issue. @pinin4fjords please advise what you prefer.

pinin4fjords commented:
Since this isn't a blocking bug fix but rather an improvement let's do 1), and it'll go in the next shinyngs release

Commit:
make hline threshold span the entire plot

Commit:
Merge branch 'single_table' of https://github.com/delfiterradas/differentialabundance into single_table

Commit:
run pre-commit run --all-files

antoniasaracco commented:
@grst @pinin4fjords I have merged the temporary solution as requested

pinin4fjords commented:
You already have my approval. Please make sure you make a pr to shinyngs with the fix so we can remove the temporary fix in future

apeltzer commented:
> You already have my approval. Please make sure you make a pr to shinyngs with the fix so we can remove the temporary fix in future

@antoniasaracco Can you please do that and add a x-ref to this PR here so that we keep it on the radar? Thank you!
